packages/actor-http-retry/lib/ActorHttpRetry.ts

```diff
@@ -0,0 +1,154 @@
+import type { IActionHttp, IActorHttpOutput, IActorHttpArgs, MediatorHttp } from '@comunica/bus-http';
+import { ActorHttp } from '@comunica/bus-http';
+import { KeysHttp } from '@comunica/context-entries';
+import { ActionContextKey, passTest, failTest } from '@comunica/core';
+import type { TestResult } from '@comunica/core';
+import type { IMediatorTypeTime } from '@comunica/mediatortype-time';
+
+export class ActorHttpRetry extends ActorHttp {
+  private readonly mediatorHttp: MediatorHttp;
+  private readonly activeDelays: Record<string, { date: Date; timeout: NodeJS.Timeout }>;
+
+  // Expression that matches dates expressed in the HTTP Date header format
+  // eslint-disable-next-line max-len
+  private static readonly dateRegex = /^(Mon|Tue|Wed|Thu|Fri|Sat|Sun), [0-9]{2} (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) [0-9]{4} [0-9]{2}:[0-9]{2}:[0-9]{2} GMT$/u;
+  private static readonly numberRegex = /^[0-9]+$/u;
+  private static readonly keyWrapped = new ActionContextKey<boolean>('urn:comunica:actor-http-retry#wrapped');
+
+  public constructor(args: IActorHttpQueueArgs) {
+    super(args);
+    this.mediatorHttp = args.mediatorHttp;
+    this.activeDelays = {};
+  }
+
+  public async test(action: IActionHttp): Promise<TestResult<IMediatorTypeTime>> {
+    if (action.context.has(ActorHttpRetry.keyWrapped)) {
+      return failTest(`${this.name} can only wrap a request once`);
+    }
+    const retryCount = action.context.get<number>(KeysHttp.httpRetryCount);
+    if (!retryCount || retryCount < 1) {
+      return failTest(`${this.name} requires a retry count greater than zero to function`);
+    }
+    return passTest({ time: 0 });
+  }
+
+  public async run(action: IActionHttp): Promise<IActorHttpOutput> {
+    const url = ActorHttp.getInputUrl(action.input);
+
+    const retryCount = action.context.getSafe<number>(KeysHttp.httpRetryCount) + 1;
+    const retryDelay = action.context.get<number>(KeysHttp.httpRetryDelay) ?? 0;
+
+    // This is declared outside the loop so it can be used for the final error message
+    let attempt = 0;
+
+    while (attempt < retryCount) {
+      attempt++;
+
+      if (url.host in this.activeDelays) {
+        this.logDebug(action.context, 'Delaying request due to host rate limit', () => ({
+          url: url.href,
+          delayedUntil: this.activeDelays[url.host].date.toISOString(),
+          attemptCount: attempt,
+        }));
+        await ActorHttpRetry.waitUntil(this.activeDelays[url.host].date);
+      }
+
+      const response = await this.mediatorHttp.mediate({
+        ...action,
+        context: action.context.set(ActorHttpRetry.keyWrapped, true),
+      });
+
+      if (response.ok) {
+        return response;
+      }
+
+      if (response.status === 429 || response.status === 503) {
```

Comment:
There may be some more 5xx range codes we may want to consider in addition.

For example, 500 and 504 could make sense as well.

I wonder if we even may want to consider making this list configurable, either through the config or the context.

https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_server_errors

Reply:
I think this was the original reason for introducing `httpRetryOnServerError`, which you removed.
I would consider keeping it, or at least a variant by which the same can be achieved.

Reply:
Retrying on 504 might make sense, not sure about 500. I can make it configurable via the context, I think that would be the cleanest solution, and would allow any response status to trigger a retry.

Reply:
I have added a new `httpRetryStatusCodes` context entry, which holds a list of status codes (as numbers), and can be used to trigger a retry for any status code and bypass the normal handling of that code (for example, to trigger a retry on 500 or even 404, or any other code that is not 'ok').

The retry actor now also retries on 504, because that one definitely does make sense.